File: clojure/sudoku/src/sudoku/io.clj
(ns sudoku.io
  (:require [clojure.string :as str])
  (:require [sudoku.layout :as layout])
  (:require [sudoku.board :as board]))

(defn read-board
  [file-path]
  (defn rowstr->intlist
    [rowstr]
    (->> (str/split rowstr #"")
         (map #(Integer/parseInt %))))
  (defn values->cell-data
    [[row colvals]]
    (-> (reduce (fn [{:keys [col data] :as acc} colval]
                  (if (> colval 0)
                    (assoc acc :col (inc col)
                               :data (conj data [[row col] colval]))
                  (assoc acc :col (inc col))))
                {:col 0
                 :data []}
                colvals)
        :data))
  (let [content (slurp file-path)
        lines (->> (str/split content #"\n")
                   (map str/trimr))
        header (first lines)
        layout_ (->> (str/split header #" ")
                     (map #(Integer/parseInt %)))
        board_ (board/make-board layout_)
        values-to-set (->> (rest lines)
                        (map rowstr->intlist)
                        (map (fn [row cvals] [row cvals])
                             (range))
                        (map values->cell-data)
                        (apply concat))]
    (reduce (fn [b [cell value]]
              (board/set-value b cell value))
            board_
            values-to-set)))

(defn print-board
  [{:keys [layout fixed] :as board}]
  (defn print-row
    [row]
    (let [positions (layout/row-cells layout row)]
      (-> (apply str
                 (map #(get fixed % \.)
                      positions))
          println)))
  (let [size (layout/size layout)]
    (dotimes [row size]
      (print-row row))))

File: js/sudoku/src/io.js
const sudoku = require('./sudoku')
const fs = require('fs')

const readSudoku = (filePath) => {

    const content = fs.readFileSync(filePath, 'utf8')
    let lines = content.split(/\r?\n/)
    const firstLine = lines.shift()
    
    const geometry = firstLine.split(/\s+/)
        .map(s => parseInt(s, 10))
    const cells = lines
        .reduce((acc, line) => acc.concat(line), "")
        .split('')
        .map(s => parseInt(s, 10))

    return {cells, geometry}

}

const printSudoku = (problem) => {

    const {cells, geometry} = problem
    const size = sudoku.getSize(problem)
    
    let row = ""
    for (let i=0; i<size*size; i++) {
        if (i % size == 0 && row.length > 0) {
            console.log(row)
            row = ""
        }
        row += cells[i] > 0 ? cells[i] : "."
    }
    console.log(row)
}

module.exports = {
    readSudoku,
    printSudoku
}